pkg/core/contracts/pools/stable/StableMath.sol

```diff
@@ -196,43 +198,42 @@ contract StableMath {
+            balanceRatiosWithFee[i] = balances[i].add(amountsIn[i]).divDown(balances[i]);
+            invariantRatioWithFees = invariantRatioWithFees.add(balanceRatiosWithFee[i].mulDown(currentWeight));
         }
 
         // Second loop calculates new amounts in, taking into account the fee on the percentage excess
         uint256[] memory newBalances = new uint256[](balances.length);
         for (uint256 i = 0; i < balances.length; i++) {
-            // Percentage of the amount supplied that will be implicitly swapped for other tokens in the pool
-            uint256 tokenBalancePercentageExcess;
-            // Some tokens might have amounts supplied in excess of a 'balanced' join: these are identified if
-            // the token's balance ratio without fee is larger than the weighted balance ratio, and swap fees are
-            // charged on the swap amount
-            if (weightedBalanceRatio >= tokenBalanceRatiosWithoutFee[i]) {
-                tokenBalancePercentageExcess = 0;
+            uint256 amountInWithoutFee;
+
+            // Check if the balance ratio is greater than the ideal ratio to charge fees or not
+            if (balanceRatiosWithFee[i] > invariantRatioWithFees) {
+                uint256 nonTaxableAmount = balances[i].mulDown(invariantRatioWithFees.sub(FixedPoint.ONE));
+                uint256 taxableAmount = amountsIn[i].sub(nonTaxableAmount);
+                amountInWithoutFee = nonTaxableAmount.add(taxableAmount.mulDown(FixedPoint.ONE.sub(swapFeePercentage)));
             } else {
-                tokenBalancePercentageExcess = tokenBalanceRatiosWithoutFee[i].sub(weightedBalanceRatio).divUp(
-                    tokenBalanceRatiosWithoutFee[i].sub(FixedPoint.ONE)
-                );
+                amountInWithoutFee = amountsIn[i];
             }
 
-            uint256 swapFeeExcess = swapFeePercentage.mulUp(tokenBalancePercentageExcess);
-
-            uint256 amountInAfterFee = amountsIn[i].mulDown(swapFeeExcess.complement());
-
-            newBalances[i] = balances[i].add(amountInAfterFee);
+            newBalances[i] = balances[i].add(amountInWithoutFee);
         }
 
-        // get the new invariant, taking swap fees into account
-        uint256 newInvariant = _calculateInvariant(amp, newBalances);
+        // Get current and new invariants, taking swap fees into account
+        uint256 currentInvariant = _calculateInvariant(amp, balances, true);
+        uint256 newInvariant = _calculateInvariant(amp, newBalances, false);
+        uint256 invariantRatio = newInvariant.divDown(currentInvariant);
 
-        // return amountBPTOut
-        return bptTotalSupply.mulDown(newInvariant.divDown(currentInvariant).sub(FixedPoint.ONE));
+        if (invariantRatio > FixedPoint.ONE) {
+            return bptTotalSupply.mulDown(invariantRatio.sub(FixedPoint.ONE));
+        } else {
+            return 0;
```

Comment:
This should only happen due to rounding, correct? Can we have a small comment explaining that if this is the case?